File: anc/anc4_mlms_distributed.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define PI           3.141592653589793
#define FS           8000.0 //mintaveteli frekvencia
#define N_S          5      //a masodlagos utakat modellezo FIR szurok hossza
#define N_W          200    //az adaptiv szurok hossza
#define MU           (1e-6) //batorsagi tenyezo
#define TMAX         3000   //ennyi utemet szimulalunk
#define COMM_PERIOD  30     //ennyi utemenkent kuldjuk az update-eket

/* a masodlagos utaknak megfelelo FIR szurok (a valosagban identifikacio eredmenye) */
double S11_id[N_S]={1,2,3,4,5};
double S12_id[N_S]={-10,9,-8,7,-6};
double S21_id[N_S]={-10,9,-8,7,-6};
double S22_id[N_S]={1,2,3,4,5};

/* a node-ok kozott megosztott valtozok (kommunikacio) */
int signal_from1to2=0;
int signal_from2to1=0;
double dW_from1to2[N_W];
double dW_from2to1[N_W];



/* tavoli zajforras */
double x(unsigned n){
    return sin(2*PI*200/FS*n);
}



/* elsodleges utak (zajforras --> mikrofon), most egyszeru IIR szurok */
double P1(double in){
    static double in1, in2, in3, out1, out2, out3;
    double out=in2+0.12*in3-0.3*out1+0.5*out3;
    in3=in2; in2=in1; in1=in;
    out3=out2; out2=out1; out1=out;
    return out;
}
double P2(double in){
    static double in1, in2, in3, out1, out2, out3;
    double out=in2+0.15*in3-0.2*out1+0.7*out3;
    in3=in2; in2=in1; in1=in;
    out3=out2; out2=out1; out1=out;
    return out;
}



/* masodlagos utak (hangszoro --> mikrofon), most a fenti Sxx_id szuroket hasznaljuk (tokeletes identifikacio) */
double S11(double in){
    static double buf[N_S];
    static unsigned i=0;
    unsigned j, ix;
    double out=0;
    buf[i]=in;
    ix=i;
    out=0;
    for(j=0;j<N_S;j++) out+=buf[ix++%N_S]*S11_id[j];
    i=(i==0)?(N_S-1):(i-1);
    return out;
}
double S12(double in){
    static double buf[N_S];
    static unsigned i=0;
    unsigned j, ix;
    double out=0;
    buf[i]=in;
    ix=i;
    out=0;
    for(j=0;j<N_S;j++) out+=buf[ix++%N_S]*S12_id[j];
    i=(i==0)?(N_S-1):(i-1);
    return out;
}
double S21(double in){
    static double buf[N_S];
    static unsigned i=0;
    unsigned j, ix;
    double out=0;
    buf[i]=in;
    ix=i;
    out=0;
    for(j=0;j<N_S;j++) out+=buf[ix++%N_S]*S21_id[j];
    i=(i==0)?(N_S-1):(i-1);
    return out;
}
double S22(double in){
    static double buf[N_S];
    static unsigned i=0;
    unsigned j, ix;
    double out=0;
    buf[i]=in;
    ix=i;
    out=0;
    for(j=0;j<N_S;j++) out+=buf[ix++%N_S]*S22_id[j];
    i=(i==0)?(N_S-1):(i-1);
    return out;
}



/* zajcsokkentes */
double node1(double x, double e){
    static double W[N_W], x_buf_W[N_W], x_buf_S[N_S], r11_buf[N_W], r21_buf[N_W], dW11[N_W], dW21[N_W];
    static unsigned iW=0, iS=0, commCntr=0;
    unsigned i, j;
    double y, r11, r21;
    /* y=-W*x */
    x_buf_W[iW]=x;
    i=iW;
    y=0;
    for(j=0;j<N_W;j++,i++) y-=x_buf_W[i%N_W]*W[j];
    /* r=S_id*x */
    x_buf_S[iS]=x;
    i=iS;
    r11=r21=0;
    for(j=0;j<N_S;j++,i++){
        r11+=x_buf_S[i%N_S]*S11_id[j];
        r21+=x_buf_S[i%N_S]*S21_id[j];
    }
    /* dW=mu*e*R */
    r11_buf[iW]=r11;
    r21_buf[iW]=r21;
    for(j=0;j<N_W;j++){
        dW11[j]+=MU*e*r11_buf[(j+iW)%N_W];
        dW21[j]+=MU*e*r21_buf[(j+iW)%N_W];
    }
    /* kommunikacio: dW21-et elkuldjuk node2-nek */
    if(commCntr==COMM_PERIOD-1){
        commCntr=0;
        for(j=0;j<N_W;j++){
            dW_from1to2[j]=dW21[j];
            dW21[j]=0;
        }
        signal_from1to2=1;
    }
    else{
        commCntr++;
    }
    /* egyutthatokeszlet frissitese, ha kaptunk update-et */
    if(signal_from2to1){
        for(j=0;j<N_W;j++){
            W[j]+=dW_from2to1[j]+dW11[j];
            dW11[j]=0;
        }
        signal_from2to1=0; //ack
    }
    /* frissites a kovetkezo utemre */
    iW=iW?(iW-1):(N_W-1);
    iS=iS?(iS-1):(N_S-1);
    return y;
}
double node2(double x, double e){
    static double W[N_W], x_buf_W[N_W], x_buf_S[N_S], r22_buf[N_W], r12_buf[N_W], dW22[N_W], dW12[N_W];
    static unsigned iW=0, iS=0, commCntr=0;
    unsigned i, j;
    double y, r22, r12;
    /* y=-W*x */
    x_buf_W[iW]=x;
    i=iW;
    y=0;
    for(j=0;j<N_W;j++,i++) y-=x_buf_W[i%N_W]*W[j];
    /* r=S_id*x */
    x_buf_S[iS]=x;
    i=iS;
    r22=r12=0;
    for(j=0;j<N_S;j++,i++){
        r22+=x_buf_S[i%N_S]*S22_id[j];
        r12+=x_buf_S[i%N_S]*S12_id[j];
    }
    /* dW=mu*e*R */
    r22_buf[iW]=r22;
    r12_buf[iW]=r12;
    for(j=0;j<N_W;j++){
        dW22[j]+=MU*e*r22_buf[(j+iW)%N_W];
        dW12[j]+=MU*e*r12_buf[(j+iW)%N_W];
    }
    /* kommunikacio: dW12-t elkuldjuk node1-nek */
    if(commCntr==COMM_PERIOD-1){
        commCntr=0;
        for(j=0;j<N_W;j++){
            dW_from2to1[j]=dW12[j];
            dW12[j]=0;
        }
        signal_from2to1=1;
    }
    else{
        commCntr++;
    }
    /* egyutthatokeszlet frissitese, ha kaptunk update-et */
    if(signal_from1to2){
        for(j=0;j<N_W;j++){
            W[j]+=dW_from1to2[j]+dW22[j];
            dW22[j]=0;
        }
        signal_from1to2=0; //ack
    }
    /* frissites a kovetkezo utemre */
    iW=iW?(iW-1):(N_W-1);
    iS=iS?(iS-1):(N_S-1);
    return y;
}



/* szimulacio */
int main(){
    unsigned long i;
    double y1, e1, y2, e2, out1=0, out2=0;
    FILE *f;
    f=fopen("signals.dat","wt");
    for(i=0;i<TMAX;i++){
        y1=P1(x(i));
        y2=P2(x(i));
        e1=y1+S11(out1)+S21(out2);
        e2=y2+S22(out2)+S12(out1);
        out1=node1(x(i),e1);
        out2=node2(x(i),e2);
        fprintf(f,"%f\t%f\t%f\t%f\n",y1,e1,y2,e2);
    }
    fclose(f);
    f=fopen("signals_plot.m","wt");
    fprintf(f,"clear all;\n");
    fprintf(f,"clc;\n");
    fprintf(f,"data=load('signals.dat');\n");
    fprintf(f,"y1=data(:,1);\n");
    fprintf(f,"e1=data(:,2);\n");
    fprintf(f,"y2=data(:,3);\n");
    fprintf(f,"e2=data(:,4);\n");
    fprintf(f,"t=(0:(size(data,1)-1))/%f;\n",FS);
    fprintf(f,"subplot(211);\n");
    fprintf(f,"plot(t,y1,t,e1,'r');\n");
    fprintf(f,"xlim([0 max(t)]);\n");
    fprintf(f,"subplot(212);\n");
    fprintf(f,"plot(t,y2,t,e2,'r');\n");
    fprintf(f,"xlim([0 max(t)]);\n");
    fclose(f);
    return 0;
}

File: anc/anc4_mlms_distributed.m
clear all;
clc;

%% PARAMETEREK
fs=8000;                    % mintaveteli frekvencia [Hz]
T=0.5;                      % szimulacios ido [s]
N_S=5;                      % masodlagos utakat modellezo FIR szurok hossza
N_W=300;                    % adaptiv szurok hossza
mu=2e-7;                    % batorsagi tenyezo
updatePeriod=[100 100];     % ennyi mintankent kuldjuk at az egyutthatokat, es frissitjuk az adaptiv szuroket
noise=@(t) sin(2*pi*200*t); % zajforras

%% ELSODLEGES UTAK (zajforras --> mikrofonok)
[PB1,PA1]=ellip(6,1,60,0.3);
[PB2,PA2]=ellip(6,1,60,0.5);

%% MASODLAGOS UTAK (Sij: i. hangszoro --> j. mikrofon)
S11=1:N_S;
S12=2*(N_S:-1:1).*(-1).^(1:N_S);
S21=S12;
S22=S11;

%% IDENTIFIKACIO (most tokeletes, a valosagban LMS algoritmus)
S_id11=S11;
S_id12=S12;
S_id21=S21;
S_id22=S22;

%% ZAJCSOKKENTES
t=0:1/fs:T;
x=noise(t);                  % zajforras (referencia)
y1=filter(PB1,PA1,x);        % zaj az 1. node-nal
y2=filter(PB2,PA2,x);        % zaj a  2. node-nal
x_shr=zeros(1,max(N_W,N_S)); % buffer a referenciajelnek
y_hat_shr=zeros(2,N_S);      % buffer a kimeneti jeleknek
r11=zeros(1,N_W);            % x megszurve S11-gyel
r12=zeros(1,N_W);            % x megszurve S12-vel
r21=zeros(1,N_W);            % x megszurve S21-gyel
r22=zeros(1,N_W);            % x megszurve S22-vel
e=zeros(2,length(t));        % hibajelek (a mikrofonok altal hallott zaj)
W=zeros(2,N_W);              % adaptiv szurok
DW=zeros(2,N_W);             % update-ek
DW_1=zeros(2,N_W);           % node1 altal (a sajat hibajele alapjan) elokeszitett update
DW_2=zeros(2,N_W);           % node2 altal (a sajat hibajele alapjan) elokeszitett update
updateCntr=zeros(1,2);       % ebben szamoljuk az update-ig

for ii=1:length(t);
    % a referenciajelet mindket node megkapja
        x_shr=[x(ii) x_shr(1:end-1)];
    % a hangszoron kiadando jel
        % node1
        y_hat=-W(1,:)*x_shr';
        y_hat_shr(1,:)=[y_hat y_hat_shr(1,1:end-1)];
        % node2
        y_hat=-W(2,:)*x_shr';
        y_hat_shr(2,:)=[y_hat y_hat_shr(2,1:end-1)];
    % a mikrofonok altal erzekelt hibajelek (ezt a "fizika csinalja")
        e(1,ii)=y1(ii)+S11*y_hat_shr(1,:)'+S21*y_hat_shr(2,:)';
        e(2,ii)=y2(ii)+S22*y_hat_shr(2,:)'+S12*y_hat_shr(1,:)';
    % update elokeszitese
        % node1
        r11=[S_id11*x_shr(1:N_S)' r11(1:end-1)];
        r21=[S_id21*x_shr(1:N_S)' r21(1:end-1)];
        DW_1(1,:)=DW_1(1,:)+r11*e(1,ii); % node1 -> node1 update
        DW_1(2,:)=DW_1(2,:)+r21*e(1,ii); % node1 -> node2 update
        % node2
        r22=[S_id22*x_shr(1:N_S)' r22(1:end-1)];
        r12=[S_id12*x_shr(1:N_S)' r12(1:end-1)];
        DW_2(2,:)=DW_2(2,:)+r22*e(2,ii); % node2 -> node2 update
        DW_2(1,:)=DW_2(1,:)+r12*e(2,ii); % node2 -> node1 update
    % kommunikacio: node1  <----DW_2(1,:)------  node2
        updateCntr=updateCntr+1;
        if updateCntr(1)>=updatePeriod(1)
            % node1: megkapja node2-tol DW_2(1,:)-et, DW_1(1,:)-et pedig o szamitja ki
            DW(1,:)=mu*DW_1(1,:)+mu*DW_2(1,:);
            W=W+DW;
            % valtozok nullazasa: csak azt a reszt, amit felhasznaltunk, a tobbit meg novelik majd a node-ok
            DW_1(1,:)=zeros(1,N_W);
            DW_2(1,:)=zeros(1,N_W);
            % teljes update matrix nullazasa 
            DW=zeros(2,N_W);
            updateCntr(1)=0;            
        end
    % kommunikacio: node1  -----DW_1(2,:)----->  node2
        if updateCntr(2)>=updatePeriod(2)
            % node2: megkapja node1-tol DW_1(2,:)-et, DW_2(2,:)-et pedig o szamitja ki
            DW(2,:)=mu*DW_1(2,:)+mu*DW_2(2,:);
            W=W+DW;
            % valtozok nullazasa: csak azt a reszt, amit felhasznaltunk, a tobbit meg novelik majd a node-ok
            DW_1(2,:)=zeros(1,N_W);
            DW_2(2,:)=zeros(1,N_W);
            % teljes update matrix nullazasa 
            DW=zeros(2,N_W);
            updateCntr(2)=0;            
        end
end

%% ABRAZOLAS
figure(1);
subplot(211);
plot(t,y1,'b',t,e(1,:),'r');
xlabel('t [s]');
ylabel('y_1, e_1');
subplot(212);
plot(t,y2,'b',t,e(2,:),'r');
xlabel('t [s]');
ylabel('y_2, e_2');